vispy/visuals/axis.py

```diff
@@ -0,0 +1,93 @@
+
+import numpy as np
+
+from .visual import Visual
+from .line import LineVisual
+from .text import TextVisual
+
+class AxisVisual(Visual):
+    """Axis visual
+
+    Parameters
+    ----------
+    extents : array
+        Co-ordinates of start and end of the axis.
+    tick_direction : array
+        Unit vector specifying the direction of the ticks. Default is
+        left/down. (This will be normalised to a unit vector)
+    domain : tuple
+        The data values at the beginning and end of the axis, used for tick
+        labels. i.e. (5, 10) means the axis starts at 5 and ends at 10.
+    scale_type : str
+        The type of scale. Default is 'linear'. Options:
+            * "linear"
+            * "logarithmic"
+            * "power"
+    """
+    def __init__(self, extents, tick_direction = None, domain=(0., 1.),
+                 scale_type="linear", axis_color=(1, 1, 1, 1),
+                 tick_color = (0.7, 0.7, 0.7, 1)):
+        self.axis_color = axis_color
+        self.tick_color = tick_color
+
+        self.extents = extents
+        self.vec = np.subtract(self.extents[1],self.extents[0])
+
+        self.tick_direction = tick_direction or self._get_tick_direction()
+
+    def draw(self, transforms):
+        v_line = LineVisual(pos=self.extents, color=self.axis_color,
+                          mode='gl', width=3.0)
+
+        tick_fractions, tick_labels = self._get_tick_frac_labels()
+
+        tick_pos, tick_label_pos = self._get_tick_positions(tick_fractions)
+
+        v_ticks = LineVisual(pos=tick_pos, color=self.tick_color, mode='gl',
+                           width=2.0, connect='segments')
+
+        # TODO: make this a more efficient iterator
+        for x in range(0, len(tick_labels)):
+            TextVisual(str(tick_labels[x]), font_size=8, color='w',
+                       pos=tick_label_pos[x]).draw(transforms)
```

Comment:
Okay. Shall I put this on hold until we've figured out what to do with
TextCollections / TextCollectionVisuals / GlyphCollections or whatever it's
being called? Or is it stable enough for me to go ahead and refactor that
all now to use the Collections?

... or should I carry on and create a fixed number of visuals and append as
needed (as @Eric89GXL suggested)?

On Mon, Mar 30, 2015 at 2:29 PM, Cyrille Rossant <notifications@github.com>
wrote:

> In vispy/visuals/axis.py
> <https://github.com/vispy/vispy/pull/829#discussion_r27389392>:
>
> > +
> > +    def draw(self, transforms):
> > +        v_line = LineVisual(pos=self.extents, color=self.axis_color,
> > +                          mode='gl', width=3.0)
> > +
> > +        tick_fractions, tick_labels = self._get_tick_frac_labels()
> > +
> > +        tick_pos, tick_label_pos = self._get_tick_positions(tick_fractions)
> > +
> > +        v_ticks = LineVisual(pos=tick_pos, color=self.tick_color, mode='gl',
> > +                           width=2.0, connect='segments')
> > +
> > +        # TODO: make this a more efficient iterator
> > +        for x in range(0, len(tick_labels)):
> > +            TextVisual(str(tick_labels[x]), font_size=8, color='w',
> > +                       pos=tick_label_pos[x]).draw(transforms)
>
> Here is how glumpy does it:
> https://github.com/glumpy/glumpy/blob/master/examples/lorenz.py
>
>
> <https://camo.githubusercontent.com/1dc47f23605a6c6ab5b3361099444e326260b1f8/687474703a2f2f676c756d70792e6769746875622e696f2f5f7374617469632f73637265656e73686f74732f6c6f72656e7a2e706e67>
>
> We have all of these collections in vispy/visuals/collections. Note that
> there are *two* visuals: one for labels, one for the grid and the ticks,
> and these are only created once. Changing them dynamically will be a bit
> tricky but you won't have to create new collections/visuals: you can just
> change the existing ones. cc @rougier <https://github.com/rougier>
>
> Personally I don't think it's worth working on a slow implementation (such
> as one where dozens of visuals are created every frame) because it will be
> far too slow to be usable
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/vispy/vispy/pull/829/files#r27389392>.
>